packages/eslint-plugin/src/rules/require-await.ts

```diff
@@ -75,13 +84,118 @@ export default createRule({
+          nodeWithAsyncKeyword,
+          token => token.value === 'async',
+        );
+
+        if (asyncToken) {
+          const nextTokenWithComments = context.sourceCode.getTokenAfter(
+            asyncToken,
+            { includeComments: true },
+          );
+
+          if (nextTokenWithComments) {
+            const asyncRange: Readonly<AST.Range> = [
+              asyncToken.range[0],
+              nextTokenWithComments.range[0],
+            ] as const;
+
+            // Removing the `async` keyword can cause parsing errors if the
+            // current statement is relying on automatic semicolon insertion.
+            // If ASI is currently being used, then we should replace the
+            // `async` keyword with a semicolon.
+            const nextToken = context.sourceCode.getTokenAfter(asyncToken);
+            const addSemiColon =
+              nextToken?.type === AST_TOKEN_TYPES.Punctuator &&
+              (nextToken.value === '[' || nextToken.value === '(') &&
+              (nodeWithAsyncKeyword.type === AST_NODE_TYPES.MethodDefinition ||
+                isStartOfExpressionStatement(nodeWithAsyncKeyword)) &&
+              needsPrecedingSemicolon(context.sourceCode, nodeWithAsyncKeyword);
+
+            const changes = [
+              { range: asyncRange, replacement: addSemiColon ? ';' : '' },
+            ];
+
+            // If there's a return type annotation and it's a
+            // `Promise<T>`, we can also change the return type
+            // annotation to just `T` as part of the suggestion.
+            // Alternatively, if the function is a generator and
+            // the return type annotation is `AsyncGenerator<T>`,
+            // then we can change it to `Generator<T>`.
+            if (node.returnType) {
+              if (scopeInfo.isGen) {
+                if (
+                  hasTypeName(node.returnType.typeAnnotation, 'AsyncGenerator')
+                ) {
+                  changes.unshift({
+                    range: node.returnType.typeAnnotation.typeName.range,
+                    replacement: 'Generator',
+                  });
+                }
+              } else if (
+                hasTypeName(node.returnType.typeAnnotation, 'Promise') &&
+                !!node.returnType.typeAnnotation.typeArguments
+              ) {
+                const openAngle = context.sourceCode.getFirstToken(
+                  node.returnType.typeAnnotation,
+                  token =>
+                    token.type === AST_TOKEN_TYPES.Punctuator &&
+                    token.value === '<',
+                );
+                const closeAngle = context.sourceCode.getLastToken(
+                  node.returnType.typeAnnotation,
+                  token =>
+                    token.type === AST_TOKEN_TYPES.Punctuator &&
+                    token.value === '>',
+                );
+                if (openAngle && closeAngle) {
+                  changes.unshift(
```

Comment:
If I change this to use `.push()`, all the test cases pass, and it's easier to read. Any reason to keep it as `unshift()`? (here and above)
```suggestion
                  changes.push(
```

Reply:
Ah, I think my assumption about how multiple edits are handled was incorrect. My thinking was that because removing the `async` keyword is always the first change, we'd want to apply this change first so that the range we specify here doesn't end up referring to a different piece of text once the `async` keyword is removed. 

Given how that doesn't seem to be the case, I'm now assuming that ESLint must adjust subsequent ranges when it applies an edit. I took a look at what ESLint does, but couldn't work out where it actually applies the changes. 😕